netrd/distance/distributional_nbd.py

```diff
@@ -0,0 +1,247 @@
+        distance_metric = self.__class__.VECTOR_DISTANCES[vector_distance]
+
+        return distance_metric(distribution_1, distribution_2)
+
+
+def shave_graph(graph):
+    """
+    Returns the two-core of a graph.
+
+    Iteratively remove the nodes of degree 0 or 1, until all nodes have
+    degree at least 2.
+
+    NOTE: duplicated from "nbd.py" to avoid excessive imports.
+
+    """
+    core = graph.copy()
+    while True:
+        to_remove = [node for node, neighbors in core.adj.items() if len(neighbors) < 2]
+        core.remove_nodes_from(to_remove)
+        if len(to_remove) == 0:
+            break
+    return core
+
+
+def pseudo_hashimoto(graph):
+    """
+    Return the pseudo-Hashimoto matrix.
+
+    The pseudo Hashimoto matrix of a graph is the block matrix defined as
+    B' = [0  D-I]
+         [-I  A ]
+
+    Where D is the degree-diagonal matrix, I is the identity matrix and A
+    is the adjacency matrix.  The eigenvalues of B' are always eigenvalues
+    of B, the non-backtracking or Hashimoto matrix.
+
+    Parameters
+    ----------
+
+    graph (nx.Graph): A NetworkX graph object.
+
+    Returns
+    -------
+
+    A sparse matrix in csr format.
+
+    NOTE: duplicated from "nbd.py" to avoid excessive imports.
+
+    """
+    # Note: the rows of nx.adjacency_matrix(graph) are in the same order as
+    # the list returned by graph.nodes().
+    degrees = graph.degree()
+    degrees = sp.diags([degrees[n] for n in graph.nodes()])
+    adj = nx.adjacency_matrix(graph)
+    ident = sp.eye(graph.order())
+    pseudo = sp.bmat([[None, degrees - ident], [-ident, adj]])
+    return pseudo.asformat('csr')
+
+
+def reduced_hashimoto(graph, shave=True, sparse=True):
+    """
+    
+
+    Parameters
+    ----------
+
+    shave (bool)
+        If True, first reduce the graph to its two-core.
+        Else graph processed in its entirety.
+
+    sparse (bool)
+        If True, returned matrix will be sparse,
+        else it will be dense.
+
+    Returns
+    -------
+
+    np.ndarray/sp.csr_matrix
+        The reduced Hashimoto Matrix.
+
+    """
+
+    if shave:
+        graph = shave_graph(graph)
+        if len(graph) == 0:
+            # We can provide a workaround for this case, however it is best
+            # that it is brought to the attention of the user.
+            raise Exception(
+                "Graph two-core is empty: non-backtracking methods unsuitable."
+            )
+
+    B = pseudo_hashimoto(graph)
+
+    if not sparse:
+        B = B.todense()
+
+    return B
+
+
+def nb_eigenvalues(B, k=None, **kwargs):
+    """
+    Calculates the eigenvalues of a matrix B.
+
+    Detects whether B is sparse/dense and uses the appropriate method.
+    If B is sparse then parameter 'k' should be provided.
+    """
+
+    if isinstance(B, np.ndarray):
+        return np.linalg.eigvals(B)
+
+    elif isinstance(B, sp.csr_matrix):
+        np.random.seed(1)  # Ensures that eigenvalue calculation is deterministic.
```

Comment:
Changing the seed in this way makes me a little uncomfortable. Would it be possible to either:

1. use `np.random.RandomState` to achieve determinism without changing the global random seed, or
2. just use a vector of 1/Ns to set up the power iteration?

Reply:
Thanks for the suggestion - I used RandomState.
See: 04cbaf9